MQClient/backends/apachepulsar.py

```diff
@@ -0,0 +1,249 @@
+        logging.debug(f"{log_msgs.ACKED_MESSAGE} ({msg_id!r}).")
+
+    def reject_message(self, msg_id: MessageID) -> None:
+        """Reject (nack) a message from the queue."""
+        if not self.consumer:
+            raise RuntimeError("queue is not connected")
+
+        logging.debug(log_msgs.NACKING_MESSAGE)
+        if isinstance(msg_id, bytes):
+            self.consumer.negative_acknowledge(pulsar.MessageId.deserialize(msg_id))
+        else:
+            self.consumer.negative_acknowledge(msg_id)
+        logging.debug(f"{log_msgs.NACKED_MESSAGE} ({msg_id!r}).")
+
+    def message_generator(self, timeout: int = 60, auto_ack: bool = True,
+                          propagate_error: bool = True) -> Generator[Optional[Message], None, None]:
+        """Yield Messages.
+
+        Generate messages with variable timeout. Close instance on exit and error.
+        Yield `None` on `throw()`.
+
+        Keyword Arguments:
+            timeout {int} -- timeout in seconds for inactivity (default: {60})
+            auto_ack {bool} -- Ack each message after successful processing (default: {True})
+            propagate_error {bool} -- should errors from downstream code kill the generator? (default: {True})
+        """
+        if not self.consumer:
+            raise RuntimeError("queue is not connected")
+
+        msg = None
+        acked = False
+        try:
+            while True:
+                # get message
+                logging.debug(log_msgs.MSGGEN_GET_NEW_MESSAGE)
+                msg = self.get_message(timeout_millis=timeout * 1000)
+                acked = False
+                if msg is None:
+                    logging.info(log_msgs.MSGGEN_NO_MESSAGE_LOOK_BACK_IN_QUEUE)
+                    break
+
+                # yield message to consumer
+                try:
+                    logging.debug(f"{log_msgs.MSGGEN_YIELDING_MESSAGE} [{msg}]")
+                    yield msg
+                # consumer throws Exception...
+                except Exception as e:  # pylint: disable=W0703
+                    logging.debug(log_msgs.MSGGEN_DOWNSTREAM_ERROR)
+                    if msg:
+                        self.reject_message(msg.msg_id)
+                    if propagate_error:
+                        logging.debug(log_msgs.MSGGEN_PROPAGATING_ERROR)
+                        raise
+                    logging.warning(f"{log_msgs.MSGGEN_EXCEPTED_DOWNSTREAM_ERROR} {e}.", exc_info=True)
+                    yield None
+                # consumer requests again, aka next()
+                else:
+                    if auto_ack:
+                        self.ack_message(msg.msg_id)
+                        acked = True
+
+        # generator exit (explicit close(), or break in consumer's loop)
+        except GeneratorExit:
+            logging.debug(log_msgs.MSGGEN_GENERATOR_EXIT)
+            if auto_ack and (not acked) and msg:
+                self.ack_message(msg.msg_id)
+                acked = True
```

Comment:
Probably don't need to update the flag here if the Generator is headed for garbage collection

Reply:
this is here in case the generator is reopened, ie not sent to the garbage collector